consensus/core/src/metrics.rs

```diff
@@ -77,24 +84,354 @@ const SIZE_BUCKETS: &[f64] = &[
+    }
+
+    pub(crate) fn load_scoring_metrics_from_storage(
+        &self,
+        metrics: Vec<(AuthorityIndex, StoredScoreMetricsU64)>,
+        hostnames: &Vec<&str>,
+    ) {
+        for ((authority, metrics), &hostname) in metrics.into_iter().zip(hostnames.iter()) {
+            let StoredScoreMetricsU64 {
+                faulty_blocks_provable_by_authority,
+                faulty_blocks_unprovable_by_authority,
+                equivocations_by_authority,
+                missing_proposals_by_authority,
+            } = metrics;
+
+            self.scoring_metrics.stored[authority.value()]
+                .faulty_blocks_provable_by_authority
+                .store(faulty_blocks_provable_by_authority, Ordering::Relaxed);
+            self.scoring_metrics.stored[authority.value()]
+                .faulty_blocks_unprovable_by_authority
+                .store(faulty_blocks_unprovable_by_authority, Ordering::Relaxed);
+            self.scoring_metrics.stored[authority.value()]
+                .equivocations_by_authority
+                .store(equivocations_by_authority, Ordering::Relaxed);
+            self.scoring_metrics.stored[authority.value()]
+                .missing_proposals_by_authority
+                .store(missing_proposals_by_authority, Ordering::Relaxed);
+
+            self.node_metrics
+                .faulty_blocks_provable_by_authority
+                .with_label_values(&[hostname, "loaded from storage", "loaded from storage"])
+                .reset();
+            self.node_metrics
+                .faulty_blocks_provable_by_authority
+                .with_label_values(&[hostname, "loaded from storage", "loaded from storage"])
+                .inc_by(faulty_blocks_provable_by_authority);
+            self.node_metrics
+                .faulty_blocks_unprovable_by_authority
+                .with_label_values(&[hostname, "loaded from storage", "loaded from storage"])
+                .reset();
+            self.node_metrics
+                .faulty_blocks_unprovable_by_authority
+                .with_label_values(&[hostname, "loaded from storage", "loaded from storage"])
+                .inc_by(faulty_blocks_unprovable_by_authority);
+            self.node_metrics
+                .uncached_equivocations_by_authority
+                .with_label_values(&[hostname])
+                .reset();
+            self.node_metrics
+                .uncached_equivocations_by_authority
+                .with_label_values(&[hostname])
+                .inc_by(equivocations_by_authority);
+            self.node_metrics
+                .uncached_missing_proposals_by_authority
+                .with_label_values(&[hostname])
+                .reset();
+            self.node_metrics
+                .uncached_missing_proposals_by_authority
+                .with_label_values(&[hostname])
+                .inc_by(missing_proposals_by_authority);
+
+            // Update score
+            self.update_score(None);
```

Comment:
Shouldn't this be Some(authority)? Otherwise you are updating all authorities score every time in this loop. I think you should argument non optional anyway, it's weird to read with the option.